### PR TITLE
[SPARK-31748][PYTHON] Document resource module in PySpark doc and rename/move classes

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR is kind of a followup for SPARK-29641 and SPARK-28234. This PR proposes:

1.. Document the new `pyspark.resource` module introduced at https://github.com/apache/spark/commit/95aec091e4d8a45e648ce84d32d912f585eeb151, in PySpark API docs.

2.. Move classes into fewer and simpler modules

Before:

```
pyspark
├── resource
│   ├── executorrequests.py
│   │   ├── class ExecutorResourceRequest 
│   │   └── class ExecutorResourceRequests
│   ├── taskrequests.py
│   │   ├── class TaskResourceRequest 
│   │   └── class TaskResourceRequests
│   ├── resourceprofilebuilder.py
│   │   └── class ResourceProfileBuilder
│   ├── resourceprofile.py
│   │   └── class ResourceProfile
└── resourceinformation.py
    └── class ResourceInformation
```

After:

```
pyspark
└── resource
    ├── requests.py
    │   ├── class ExecutorResourceRequest 
    │   ├── class ExecutorResourceRequests
    │   ├── class TaskResourceRequest 
    │   └── class TaskResourceRequests
    ├── profile.py
    │   ├── class ResourceProfileBuilder
    │   └── class ResourceProfile
    └── information.py
        └── class ResourceInformation
```

3.. Minor docstring fix e.g.:

```diff
-     @param name the name of the resource
-     @param addresses an array of strings describing the addresses of the resource
+     :param name: the name of the resource
+     :param addresses: an array of strings describing the addresses of the resource
+
+     .. versionadded:: 3.0.0
```

### Why are the changes needed?

To document APIs, and move Python modules to fewer and simpler modules.

### Does this PR introduce _any_ user-facing change?

No, the changes are in unreleased branches.

### How was this patch tested?

Manually tested via:

```bash
cd python
./run-tests --python-executables=python3 --modules=pyspark-core
./run-tests --python-executables=python3 --modules=pyspark-resource
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @holdenk too - seems I can't add you as a reviewer.

File: python/pyspark/resource/information.py [hunk 1]
```diff
@@ -26,8 +26,10 @@ class ResourceInformation(object):
 
     One example is GPUs, where the addresses would be the indices of the GPUs
 
-    @param name the name of the resource
-    @param addresses an array of strings describing the addresses of the resource
+    :param name: the name of the resource
+    :param addresses: an array of strings describing the addresses of the resource
+
+    .. versionadded:: 3.0.0
```
Comment (HyukjinKwon) [inline python/pyspark/resource/information.py:32]: Will add a PR to backport this change specifically for branch-3.0. This PR itself should go to master only.

**Review by BryanCutler — APPROVED** [review]
+1 on the new module layout and the doc changes LGTM

**Review by viirya — APPROVED** [review]
Looks good too. It looks clearer.

File: python/pyspark/resource/profile.py [hunk 2]
```diff
@@ -15,10 +15,61 @@
...
+
+class ResourceProfile(object):
+
+    """
+    .. note:: Evolving
+
+    Resource profile to associate with an RDD. A :class:`pyspark.resource.ResourceProfile`
+    allows the user to specify executor and task requirements for an RDD that will get
+    applied during a stage. This allows the user to change the resource requirements between
+    stages. This is meant to be immutable so user doesn't change it after building.
```
Comment (viirya) [inline python/pyspark/resource/profile.py:30]: nit: doesn't change -> cannot change?

File: python/pyspark/resource/requests.py [hunk 3]
```diff
@@ -167,3 +166,89 @@ def requests(self):
...
+
+
+class TaskResourceRequest(object):
+    """
+    .. note:: Evolving
+
+    A task resource request. This is used in conjuntion with the
+    :class:`pyspark.resource.ResourceProfile` to programmatically specify the resources
+    needed for an RDD that will be applied at the stage level. The amount is specified
+    as a Double to allow for saying you want more then 1 task per resource. Valid values
```
Comment (viirya) [inline python/pyspark/resource/requests.py:178]: typo: more then -> more than

**Comment by viirya** [comment]
retest this please

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master. Thank you.

File: python/pyspark/resource/information.py (hunk 1 above)
Comment (dongjoon-hyun) [reply python/pyspark/resource/information.py:32]: Please make a PR to branch-3.0 for this part.

**Comment by HyukjinKwon** [comment]
Thank you all!

**Pull request closed without merging**